### PR TITLE
Update project outputs

### PR DESCRIPTION
**Pull request by odeimaiz**
## What do these changes do?

- [x] transfer progress of computation services in the project db table @sanderegg 
- [x] transfer output payload from comp_task table to project db table, to allow frontend to update and show the outputs of each individual node in the workbench @sanderegg 
- [x] update frontend accordingly @odeimaiz 

## Related issue number
#1260 

updateOutputs

## Checklist

- [ ] Did you change any service's API? Then make sure to bundle document and upgrade version (``make openapi-specs``, ``git commit ...`` and then ``make version-*``)
- [ ] Unit tests for the changes exist
- [ ] Runs in the swarm
- [ ] Documentation reflects the changes
- [ ] New module? Add your github username to [.github/CODEOWNERS](.github/CODEOWNERS)

### CODE REVIEW

File: services/web/server/src/simcore_service_webserver/computation_comp_tasks_listening_task.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+from .projects.projects_models import projects, user_to_projects
+
+log = logging.getLogger(__name__)
+
+DB_PROCEDURE_NAME: str = "notify_comp_tasks_changed"
+DB_TRIGGER_NAME: str = f"{DB_PROCEDURE_NAME}_event"
+DB_CHANNEL_NAME: str = "comp_tasks_output_events"
+async def register_trigger_function(app: web.Application):
+    db_engine: Engine = app[APP_DB_ENGINE_KEY]
+    # NOTE: an example was found in https://citizen428.net/blog/asynchronous-notifications-in-postgres/
```
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/computation_comp_tasks_listening_task.py]: Check how to do this directly iwht aiopg https://aiopg.readthedocs.io/en/stable/examples.html#usage-of-listen-notify-commands

File: services/web/server/src/simcore_service_webserver/computation_comp_tasks_listening_task.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+from .computation_config import APP_COMP_TASKS_LISTENING_KEY
+from .projects import projects_api
+from .projects.projects_models import projects, user_to_projects
+
+log = logging.getLogger(__name__)
+
+DB_PROCEDURE_NAME: str = "notify_comp_tasks_changed"
+DB_TRIGGER_NAME: str = f"{DB_PROCEDURE_NAME}_event"
+DB_CHANNEL_NAME: str = "comp_tasks_output_events"
```
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/computation_comp_tasks_listening_task.py]: MINOR: leave some space between const and functions

File: services/web/server/src/simcore_service_webserver/computation_comp_tasks_listening_task.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    trigger_registration_query = f"""
+    CREATE TRIGGER {DB_TRIGGER_NAME}
+    AFTER UPDATE OF outputs ON comp_tasks
+        FOR EACH ROW
+        WHEN (OLD.outputs::jsonb IS DISTINCT FROM NEW.outputs::jsonb)
+        EXECUTE PROCEDURE {DB_PROCEDURE_NAME}();
+    """
+    async with db_engine.acquire() as conn:
+        await conn.execute(notification_fct_query)
+        await conn.execute(trigger_registration_query)
```
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/computation_comp_tasks_listening_task.py]: use [``conn.begin()``](https://aiopg.readthedocs.io/en/stable/sa.html#aiopg.sa.SAConnection.begin) to rollback transactions

see [examples](https://aiopg.readthedocs.io/en/stable/examples.html#simple-transaction-in-sqlalchemy)

File: services/web/server/src/simcore_service_webserver/computation_comp_tasks_listening_task.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        log.info("listening to comp_task events...")
+        while True:
+            await listen(app)
+    except asyncio.CancelledError:
+        pass
+    finally:
+        await unregister(app)
+
+async def setup_comp_tasks_listening_task(app: web.Application):
+    app[APP_COMP_TASKS_LISTENING_KEY] = asyncio.get_event_loop().create_task(comp_tasks_listening_task(app))
```
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/computation_comp_tasks_listening_task.py]: if you do not need to use ``app[APP_COMP_TASKS_LISTENING_KEY]``, why do you store it?
can just use ``task`` variable and use it within the context after the ``yield`` as well

File: services/web/server/src/simcore_service_webserver/computation_comp_tasks_listening_task.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+async def unregister(app: web.Application) -> None:
+    # unlisten_query = f"UNLISTEN {DB_CHANNEL_NAME}"
+    drop_trigger_query = f"""
+    DROP TRIGGER {DB_TRIGGER_NAME} on comp_tasks;
+    """
+
+    db_engine: Engine = app[APP_DB_ENGINE_KEY]
+    async with db_engine.acquire() as conn:
+        # await conn.execute(unlisten_query)
```
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/computation_comp_tasks_listening_task.py]: MINOR: remove unused code or add note pls

File: services/web/server/src/simcore_service_webserver/computation_subscribe.py [hunk 6]
```diff
@@ -2,56 +2,86 @@
...
         @wraps(func)
-        async def wrapped(*args, **kwargs):
+        async def wrapped(*args, **kwargs) -> Coroutine:
             return await func(*args, **kwargs, app=app)
         return wrapped
     return decorator
 
 
-async def on_message(message: aio_pika.IncomingMessage, app: web.Application):
+async def post_messages(app: web.Application, user_id: str, messages: Dict[str, Any]) -> None:
```
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/computation_subscribe.py]: MINOR: Improve clarity. I find confusing the name ``post_message`` in a module that has code for rabbit ( a message broker) and websockets (that exchange messages with the front-end).

File: services/web/server/src/simcore_service_webserver/computation_comp_tasks_listening_task.py (hunk 1 above)
Comment (sanderegg) [reply services/web/server/src/simcore_service_webserver/computation_comp_tasks_listening_task.py]: Yes that is what is done (see line 76) with the listen function.
As is shown in [this link](https://citizen428.net/blog/asynchronous-notifications-in-postgres/) it is not sufficient to just "listen" to the events. The database needs some programming as well, which is what **register_trigger_function** does. A fct needs to be registered within the database + a TRIGGER needs to be defined [see here](https://www.postgresql.org/docs/9.1/sql-createtrigger.html).
This should be created somehow through alembic I guess?

File: services/web/server/src/simcore_service_webserver/computation_comp_tasks_listening_task.py (hunk 3 above)
Comment (sanderegg) [reply services/web/server/src/simcore_service_webserver/computation_comp_tasks_listening_task.py]: done

File: services/web/server/src/simcore_service_webserver/computation_comp_tasks_listening_task.py (hunk 4 above)
Comment (sanderegg) [reply services/web/server/src/simcore_service_webserver/computation_comp_tasks_listening_task.py]: well I agree.. this comes from a copy/paste of mine where I was using it to check for the task state. I'll remove it.

File: services/web/client/source/class/osparc/component/widget/inputs/NodeOutputTreeItem.js [hunk 7]
```diff
@@ -109,6 +110,11 @@ qx.Class.define("osparc.component.widget.inputs.NodeOutputTreeItem", {
     _applyValue: function(value) {
       this.__valueLabel.setValue(value);
     },
+
+    __isNumber: function(n) {
```
Comment (ignapas) [inline services/web/client/source/class/osparc/component/widget/inputs/NodeOutputTreeItem.js]: that code is weird!

File: services/web/client/source/class/osparc/data/model/Node.js [hunk 8]
```diff
@@ -514,7 +518,11 @@ qx.Class.define("osparc.data.model.Node", {
     setOutputData: function(nodeData) {
       if (nodeData.outputs) {
         for (const outputKey in nodeData.outputs) {
-          this.__outputs[outputKey].value = nodeData.outputs[outputKey];
+          if (!Object.prototype.hasOwnProperty.call(this.__outputs, outputKey)) {
+            this.__outputs[outputKey] = {};
+          }
+          this.__outputs[outputKey]["value"] = nodeData.outputs[outputKey];
```
Comment (ignapas) [inline services/web/client/source/class/osparc/data/model/Node.js]:
```suggestion
          this.__outputs[outputKey].value = nodeData.outputs[outputKey];
```

File: services/web/client/source/class/osparc/desktop/StudyEditor.js [hunk 9]
```diff
@@ -474,6 +474,24 @@ qx.Class.define("osparc.desktop.StudyEditor", {
         }
       }, this);
 
+      // callback for node updates
+      const slotName3 = "nodeUpdated";
+      socket.removeSlot(slotName3);
```
Comment (ignapas) [inline services/web/client/source/class/osparc/desktop/StudyEditor.js]: cool... but what is that ``removeSlot`` for?

File: services/web/client/source/class/osparc/data/model/Node.js (hunk 8 above)
Comment (odeimaiz) [reply services/web/client/source/class/osparc/data/model/Node.js]: Why?

File: services/web/client/source/class/osparc/desktop/StudyEditor.js (hunk 9 above)
Comment (odeimaiz) [reply services/web/client/source/class/osparc/desktop/StudyEditor.js]: It remove the listener if it already exists. Without it, if you start the same pipeline twice, you also subscribe to that message two times.

File: services/web/server/src/simcore_service_webserver/computation_subscribe.py (hunk 6 above)
Comment (sanderegg) [reply services/web/server/src/simcore_service_webserver/computation_subscribe.py]: ok I separated the modules. it should be clearer and PRed to Odeis PR.

**Comment by sanderegg** [comment]
> Looks very cool, but why has to be the DB notifying the changes... shouldn't it be the guy creating those outputs?

this could be done but not with the current solution where dynamic services are just pushing to the DB at any time. when the new sidecar comes in this could be re-discussed.

**Pull request merged**